From 4af77e8208f45d045fc0848704e50db53b3c0ba4 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Thu, 16 Mar 2017 14:35:19 +0800
Subject: [PATCH 0904/1345] u3d: Add PM support for USB3.0 device driver

commit  9c8785b6c4d21f823ac14820c7afe2515145876d from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

- Since U3D and UDC share the same device controller, need to
  suspend/resume together with the UDC driver
- Stop the phy driver in suspend, start again in resume
- Stop all USB activities in suspend, start again in resume

Change-Id: I0c132ee2b2806f1d1e4bc9e98f8075d1f9190563
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37506
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/gadget/udc/mvebu_u3d.c |   33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/drivers/usb/gadget/udc/mvebu_u3d.c b/drivers/usb/gadget/udc/mvebu_u3d.c
index c768c7c..2c46a4b 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.c
+++ b/drivers/usb/gadget/udc/mvebu_u3d.c
@@ -2458,11 +2458,44 @@ static int mvc2_probe(struct platform_device *pdev)
 #ifdef CONFIG_PM
 static int mvc2_suspend(struct device *dev)
 {
+	struct mvc2 *cp = (struct mvc2 *)dev_get_drvdata(dev);
+
+	/* Stop the current activities */
+	if (cp->driver)
+		stop_activity(cp, cp->driver);
+
+	/* PHY exit if there is */
+	if (cp->comphy) {
+		phy_power_off(cp->comphy);
+		phy_exit(cp->comphy);
+	}
+
 	return 0;
 }
 
 static int mvc2_resume(struct device *dev)
 {
+	struct mvc2 *cp = (struct mvc2 *)dev_get_drvdata(dev);
+	int ret;
+
+	/* PHY init if there is */
+	if (cp->comphy) {
+		ret = phy_init(cp->comphy);
+		if (ret)
+			return ret;
+
+		ret = phy_power_on(cp->comphy);
+		if (ret) {
+			phy_power_off(cp->comphy);
+			phy_exit(cp->comphy);
+			return ret;
+		}
+	}
+
+	/* Start the current device if driver is connected */
+	if (cp->driver)
+		mvc2_start(&cp->gadget, cp->driver);
+
 	return 0;
 }
 
-- 
1.7.9.5

